File: Recursion/12. Tower Of Hanoi - GFG/12.-tower-of-hanoi.java
// { Driver Code Starts
// Initial Template for Java

/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;
class Recursion {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int T = sc.nextInt();//total testcases
        while (T-- > 0) {
            Hanoi obj = new Hanoi();
            int N;
            
            //taking input N
            N = sc.nextInt();

            //calling toh() method 
            System.out.println(obj.toh(N, 1, 3, 2));
        }
    }
}





// } Driver Code Ends


// User function Template for Java


// avoid space at the starting of the string in "move disk....."
class Hanoi {

    public long toh(int N, int from, int to, int aux) {
        // Your code here
        if (N==1){ //base case
            System.out.println("move disk " +N + " from rod "+ from + " to rod " + to);
            return 1;
        }
        
        // recursive call to move to aux
        toh(N-1,from,aux,to);
        System.out.println("move disk " +N + " from rod " + from + " to rod " + to);
        
        // recusive call to move to destination
        toh(N-1,aux,to,from);
        return (long)Math.pow(2,N)-1;
        
    }
}

File: Recursion/12. Tower Of Hanoi - GFG/12.-tower-of-hanoi.cpp
// { Driver Code Starts
#include <bits/stdc++.h>

using namespace std;


 // } Driver Code Ends
class Solution{
    public:
    // You need to complete this function

    // avoid space at the starting of the string in "move disk....."
    long long toh(int N, int from, int to, int aux) {
        // Your code here
    long long count = 1;
    
    if(N==1){        //base case
        cout << "move disk 1 from rod " << from << " to rod " << to << endl;
        return 1;
    }
    
    // move smaller to auxillary
    count += toh(N-1,from,aux,to);
    cout << "move disk " << N << " from rod " << from << " to rod " << to << endl;
    
    // move bigger to destination
    count += toh(N-1,aux,to,from);
    return count;
    
  }

};

// { Driver Code Starts.

int main() {

    int T;
    cin >> T;//testcases
    while (T--) {
        
        int N;
        cin >> N;//taking input N
        
        //calling toh() function
        Solution ob;
        
        cout << ob.toh(N, 1, 3, 2) << endl;
    }
    return 0;
}


//Position this line where user code will be pasted.  // } Driver Code Ends